### ISSUE TITLE
Cabal generated packages version string for use by the package.

### ISSUE THREAD

**Issue by erikd** [haskell/cabal#3233]
I think it would be nice to have a cabal generated version string that the package could then use, similar to the way cabal generates a `Path_${pkgname}` module containing things like `bindir`, `libdir` and `datadir` etc.

I can see two uses for this functionality:

* For executables: every executable should be able to print out its version number. Obviously this can be hard coded in a `String` that gets updated for every release, but thats a pain in the neck. It would be much better to have the correct version string autogenerated from the version number in the package's cabal file.

* For libraries: a library can export its version string, so that client programs that use that library can print out the version of the library the client program is using.

If I was to implement thing, I would have cabal generate a file `dist/build/autogen/Version_$(pkgname)` containing a single export:

```
module Version_$(pkgname)
versionString :: String
versionString = "w.x.y.z
```

My use case is for the `hjsmin` program which depends on `language-javascript`. If this cabal feature were available, `hjsmin` would be able to do the following:

```
$ hjsmin --version
hjsmin version 1.2.3 using language-javascript version 3.4.5
```

If people think this is a good idea and like the way I am suggesting to implement it, I will whip up some patches.

PS: Before logging this issue I trawled through all 25 pages of issues here in the cabal github issue tracker and couldn't anything like this.

**Comment by 23Skidoo** [comment]
`cabal_macros.h` defines `VERSION_foo` macros for all dependencies. So you should be already able to tell which version of `language-javascript` your program was compiled against.

**Comment by 23Skidoo** [comment]
Related: #1506.

**Comment by erikd** [comment]
The `Version_foo` macros are very useful. Its a pity I have to polute my beautfile Haskell code with `{-# LANGUAGE CPP #-}` and `#include "cabal_macros.h"` :) Also, it seems that `cabal_macros.h` does not include the version string for the package itself so maybe thats a first step.

**Comment by 23Skidoo** [comment]
You don't need to `#include` this file, it is done automatically.

>  Also, it seems that `cabal_macros.h` does not include the version string for the package itself

Should be easy enough to add this.

**Comment by erikd** [comment]
Related : https://github.com/haskell/cabal/pull/3235

**Comment by 23Skidoo** [comment]
Fixed by #3235.
